src/realm/array_decimal128.cpp

```diff
@@ -21,40 +21,127 @@
 
 namespace realm {
 
+namespace {
+
+uint8_t min_width(const Decimal128& value)
+{
+    Decimal128::Bid128 coefficient;
+    int exponent;
+    bool sign;
+
+    if (value.is_null()) {
+        return 0;
+    }
+
+    value.unpack(coefficient, exponent, sign);
+    if (coefficient.w[1] == 0) {
+        if (coefficient.w[0] < (1ull << 23) && exponent > -95 && exponent < 96) {
```

Comment:
it would be ideal to have some tests explicitly checking upgrades around these value boundaries

Reply:
Good idea! There were actually some problems.